core/src/main/java/com/linecorp/armeria/client/ConnectionPoolMetrics.java

```diff
@@ -15,50 +15,129 @@
+    private final ReentrantShortLock lock = new ReentrantShortLock();
 
     /**
      * Creates a new instance with the specified {@link Meter} name.
      */
     ConnectionPoolMetrics(MeterRegistry meterRegistry, MeterIdPrefix idPrefix) {
-        this.idPrefix = requireNonNull(idPrefix, "idPrefix");
-        this.meterRegistry = requireNonNull(meterRegistry, "meterRegistry");
+        this.idPrefix = idPrefix;
+        this.meterRegistry = meterRegistry;
     }
 
     void increaseConnOpened(SessionProtocol protocol, InetSocketAddress remoteAddr,
                             InetSocketAddress localAddr) {
-        meterRegistry.counter(idPrefix.name(),
-                              idPrefix.tags(PROTOCOL, protocol.name(),
-                                            REMOTE_IP, remoteAddr.getAddress().getHostAddress(),
-                                            LOCAL_IP, localAddr.getAddress().getHostAddress(),
-                                            STATE, "opened"))
-                     .increment();
+        final List<Tag> commonTags = commonTags(protocol, remoteAddr, localAddr);
+        lock.lock();
+        try {
+            final Meters meters = metersMap.computeIfAbsent(commonTags,
+                                                            key -> new Meters(idPrefix, key, meterRegistry));
+            meters.increment();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private List<Tag> commonTags(SessionProtocol protocol, InetSocketAddress remoteAddr,
+                                 InetSocketAddress localAddr) {
+        return idPrefix.tags(PROTOCOL, protocol.name(),
+                             REMOTE_IP, remoteAddr.getAddress().getHostAddress(),
+                             LOCAL_IP, localAddr.getAddress().getHostAddress());
     }
 
     void increaseConnClosed(SessionProtocol protocol, InetSocketAddress remoteAddr,
                             InetSocketAddress localAddr) {
-        meterRegistry.counter(idPrefix.name(),
-                              idPrefix.tags(
-                                      PROTOCOL, protocol.name(),
-                                      REMOTE_IP, remoteAddr.getAddress().getHostAddress(),
-                                      LOCAL_IP, localAddr.getAddress().getHostAddress(),
-                                      STATE, "closed"))
-                     .increment();
+
+        final List<Tag> commonTags = commonTags(protocol, remoteAddr, localAddr);
+        lock.lock();
+        try {
+            final Meters meters = metersMap.get(commonTags);
+            if (meters != null) {
+                meters.decrement();
+                assert meters.activeConnections() >= 0 : "active connections should not be negative" + meters;
+                if (meters.activeConnections() == 0) {
+                    // XXX(ikhoon): Should we consider to remove the gauge lazily so that collectors can get the
```

Comment:
I'm not sure what this comment means

Reply:
I've never actually tried with Prometheus but it seems Prometheus keeps a value as is if the value is absent from the metrics. As we remove the meters from `MeterRegistry` Prometheus may show the old values for 5 minutes.
https://stackoverflow.com/questions/56072858/treat-value-as-0-if-not-update-in-prometheus
Thus, I thought, is it worth scheduling a task to remove a stale metric after 10 minutes?

Reply:
Do you mean the prometheus collector implementation in java purges missing data?  (not the actual backend storage)

Reply:
No, it meant the backend storage behavior. The meters whose number of active connections is zero will be removed from `MeterRegistry`.
https://github.com/line/armeria/blob/55b0c21107be777e3eb249a259af180a90cbdc89/core/src/test/java/com/linecorp/armeria/client/ConnectionPoolCollectingMetricTest.java#L64-L72
So the meter for zero active connection is not `armeria.client.connections#value{...}=0` but null.

I was concerned that dashboard for the metrics would not reflect a zero value. I think the value will be stale (1) for 5 minutes even if the value is actually zero.

Reply:
Then, probably better to use `Caffeine`? If so, we could also limit the size or use weak key and value.

Reply:
Commented on https://github.com/line/armeria/pull/5288#issuecomment-1833270827